Sources/SwiftLexicalLookup/Configurations/FileScopeNameIntroductionStrategy.swift

```diff
@@ -0,0 +1,25 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+/// Specifies how names should be introduced at the file scope.
+@_spi(Experimental) public enum FileScopeHandlingConfig {
+  /// Default behavior. Names introduced sequentially like in member block
+  /// scope up to the first non-declaration after and including which,
+  /// the declarations are treated like in code block scope.
+  case memberBlockUpToLastDecl
+  /// File scope behaves like member block scope.
+  case memberBlock
+  /// File scope behaves like code block scope.
+  case codeBlock
```

Comment:
We don’t make any assumptions about the kind of top-level behavior other than defaulting to `memberBlockUpToLastDecl ` when not specified explicitly by the client. Do you think we should add code snippets describing lookup behavior for each case? I already added some examples [here](https://github.com/swiftlang/swift-syntax/blob/940b06e4d8006286de1f0ba89d52b27e1368dad7/Sources/SwiftLexicalLookup/ScopeImplementations.swift#L83-L102). Do you think we should add comments suggesting when clients should use each of the behaviors?

Reply:
I was thinking of little sentences like the following (please check if they are actually correct):
- `memberBlock`: This is the behavior that is being used for Swift files that don’t allow top-level code
- `memberBlockUpToLastDecl`: This is the behavior that is being used for Swift files with top-level code
- `codeBlock`: I don’t actually know what this is for. Or is this the behavior used for top-level code?

Reply:
I thought `codeBlock` could be used e.g. when wanting to apply the lookup only to extracted function, loop, if statement body etc. I also added it for the sake of completion, to cover all possible cases. I added suggested comments.

Reply:
How hard would it be to split the changes that address my review comments here from the changes for `guard` lookup in https://github.com/swiftlang/swift-syntax/pull/2748? That would make it easier to review and would allow us to eg. get swift-syntax building with Swift 5.8 without needing to wait for  the changes for `guard` lookup.

Reply:
I created #2755 with the requested changes and without `guard` and implicit name lookup. I hope I haven't missed anything.

Reply:
Thank you!